dallinger/docker/tools.py

```diff
@@ -0,0 +1,188 @@
+            env={**os.environ.copy(), **env},
+        )
+        check_output("docker-compose up -d".split(), env={**os.environ.copy(), **env})
+        # Wait for postgres to complete initialization
+        while b"ready to accept connections" not in check_output(
+            [
+                "docker-compose",
+                "-f",
+                f"{self.tmp_dir}/docker-compose.yml",
+                "logs",
+                "postgresql",
+            ]
+        ):
+            time.sleep(2)
+        try:
+            check_output(
+                [
+                    "docker-compose",
+                    "-f",
+                    f"{self.tmp_dir}/docker-compose.yml",
+                    "exec",
+                    "worker",
+                    "dallinger-housekeeper",
+                    "initdb",
+                ]
+            )
+        except CalledProcessError:
+            self.out.error("There was a problem initializing the database")
+            self.stop()
+            raise
+        # Make sure the containers are all started
+        status = check_output(
+            [
+                "docker-compose",
+                "-f",
+                f"{self.tmp_dir}/docker-compose.yml",
+                "ps",
+            ]
+        )
+        errors = []
+        # docker-compose output looks like this:
+        #             Name                           Command               State     Ports
+        # -----------------------------------------------------------------------------------
+        # function_learning_postgresql_1   docker-entrypoint.sh postgres    Up       5432/tcp
+        # function_learning_redis_1        docker-entrypoint.sh redis ...   Up       6379/tcp
+        # function_learning_web_1          /bin/sh -c dallinger_herok ...   Exit 1
+        # function_learning_worker_1       /bin/sh -c dallinger_herok ...   Up
+        # ^^^ a final newline
+        for line in status.decode("utf-8").split("\n")[2:-1]:
+            if "Up" not in line:
+                errors.append(line)
+        if errors:
+            self.out.error("Some services did not start properly:")
+            for error in errors:
+                self.out.error(error)
+                self.out.error(
+                    self.client.attach(error.split(" ")[0], logs=True).decode("utf-8")
+                )
+            raise DockerStartupError
+        return self
+
+    def __exit__(self, exctype, excinst, exctb):
+        self.stop()
+
+    def stop(self):
+        os.system(f"docker-compose -f '{self.tmp_dir}/docker-compose.yml' down")
+
+    def get_container_name(self, service_name):
+        """Return the name of the first container for the given service name
+        as it is known to docker, as opposed to docker-compose.
+        """
+        return f"{self.experiment_name}_{service_name}_1"
+
+    def monitor(self, listener):
+        # How can we get a stream for two containers?
```

Comment:
I don't think we've ever worried about the worker log stream for this purpose, so this should be fine.

Reply:
For me it was painful because I had errors in the worker and I needed to fish them out with (relatively) arcane `docker-compose` invocations. I believe we should talk a bit more about this.